### PR TITLE
Add nvidia smi pci bandwidth percent collector

### PR DESCRIPTION
**Pull request by ghanapunq**
##### Summary
This change enables the calculation of the PCIE bandwidth in percent.  

##### Test Plan

Tested manually on two different machines. One with a GPU Quadro RTX 4000 and the other one with GPU RTX  A4500
##### Additional Information

To have a quick overview about the pcie bandwidth available, I though it was useful to have the values of TX/RX bandwidth in percent. To do so it calculates the maximum theoretical bandwidth thanks to the pcie generation and width given by nvidia-smi, thus the value in percent can be deduced.

<details> <summary>For users: metric netdata_nvidia_smi_pci_bandwidth_KiB_persec_average enhanced  </summary>

It changes the nvidia-smi plugin only. And two dimensions _rx_percent_ and _tx_percent_ have been added to the _netdata_nvidia_smi_pci_bandwidth_KiB_persec_average_ metric.

</details>

### CODE REVIEW

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py [hunk 1]
```diff
@@ -147,6 +151,12 @@ def gpu_charts(gpu):
                 ['user_num', 'users'],
             ]
         },
+        PCI_GEN: {
+            'options': [None, 'PCI Generation', 'num', fam, 'nvidia_smi.pci_gen', 'line'],
+            'lines': [
+                ['pci_link_gen', 'gen', 'absolute', 1, 1],
+             ]
+        },
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:159]: Why do we need `pci_link_gen` as a chart/metric?

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py [hunk 2]
```diff
@@ -326,6 +336,20 @@ def name(self):
     def full_name(self):
         return 'gpu{0} {1}'.format(self.num, self.name())
 
+    @handle_attr_error
+    def pci_link_gen(self):
+        return self.root.find('pci').find('pci_gpu_link_info').find('pcie_gen').find('max_link_gen').text
+
+    @handle_attr_error
+    def pci_link_width(self):
+        return self.root.find('pci').find('pci_gpu_link_info').find('link_widths').find('max_link_width').text.split('x')[0]
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:345]: Are you sure we need to use max and not current?

> **Current**
> The current link generation and width. These may be reduced when the GPU is not in use.

> **Maximum**
> The maximum link generation and width possible with this GPU and system configuration. For example, if the GPU supports a higher PCIe generation than the system.

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py [hunk 3]
```diff
@@ -326,6 +336,20 @@ def name(self):
...
+
+    def pci_bw_max(self):
+        link_gen = int(self.pci_link_gen())
+        link_width = int(self.pci_link_width())
+        # return max bandwidth in kB/s
+        return link_gen * 250 * 1024 * link_width
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:351]: Can you explain (or share a link to the doc) this calculation? What is 250?

I [googled](https://enterprise-support.nvidia.com/s/article/understanding-pcie-configuration-for-maximum-performance#:~:text=Maximum%20PCIe%20Bandwidth%20%3D%20SPEED%20*%20WIDTH,1G%20%3D%20~62Gb%2Fs.):

> The maximum possible PCIe bandwidth is calculated by multiplying the PCIe width and speed. From that number, we reduce ~1Gb/s for error correction protocols and the PCIe headers overhead. The overhead is determined by both the PCIe encoding (see PCIe speed for details), and the PCIe MTU:

> **Maximum PCIe Bandwidth = SPEED * WIDTH * (1 - ENCODING) - 1Gb/s**.

Comment (ghanapunq) [reply collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:351]: Indeed the calculation is not right, I need to rework it.

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py (hunk 2 above)
Comment (ghanapunq) [reply collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:345]: Yes, I want the maximum capable bandwidth which is constant whereas the current is meant to evolve accordingly to the GPU consumption.

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py [hunk 4]
```diff
@@ -326,6 +332,22 @@ def name(self):
...
+    def pci_link_width(self):
+        return self.root.find('pci').find('pci_gpu_link_info').find('link_widths').find('max_link_width').text.split('x')[0]
+
+    def pci_bw_max(self):
+        link_gen = int(self.pci_link_gen())
+        link_width = int(self.pci_link_width())
+        # Maximum PCIe Bandwidth = SPEED * WIDTH * (1 - ENCODING) - 1Gb/s.
+        # see details https://enterprise-support.nvidia.com/s/article/understanding-pcie-configuration-for-maximum-performance
+        # return max bandwidth in kilobytes per second (kB/s)
+        return (PCI_SPEED[link_gen] * link_width * (1- PCI_ENCODING[link_gen]) - 1) * 1000 * 1000 / 8
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:349]: > PCI_SPEED = [2.5, 5, 8, 16, 32]
> PCI_ENCODING = [2/10, 2/10, 2/130, 2/130, 2/130]
> PCI_SPEED[link_gen]
> PCI_ENCODING[link_gen])

I think this is wrong

```xml
<pcie_gen>
	<max_link_gen>3</max_link_gen>
</pcie_gen>
```

> > where 2.5GT/s is referred as "gen1", 5GT/s as "gen2", and 8GT/s as "gen3"

PCI_SPEED[3] == 16, but should be 8.

You need to either use `link_gen - 1` as an index or make `PCI_SPEED` a dictionary (link_gen: speed).

```python
PCI_SPEED = {
  "1": 2.5, 
  "2": 5, 
  "3": 8, 
  "4": 16,
  "5": 32
}
```

**Comment by ilyam8** [comment]
@ghanapunq also check the [Codacy warnings](https://app.codacy.com/gh/netdata/netdata/pullRequest?prid=10996343)

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py [hunk 5]
```diff
@@ -474,6 +481,14 @@ def data(self):
             'power_draw': self.power_draw(),
         }
 
+        if self.pci_bw_max() is None or self.pci_bw_max() == 0:
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:484]:
```suggestion
        pci_bw_max = self.pci_bw_max()
        if not pci_bw_max:
            ...
```

nit-pick:
  - call `self.pci_bw_max()` once, later use `pci_bw_max`.
  - use `not pci_bw_max` instead `is None or == 0` (does the same job).

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py [hunk 6]
```diff
@@ -439,6 +481,16 @@ def data(self):
             'power_draw': self.power_draw(),
         }
 
+        pci_bw_max = self.pci_bw_max()
+        if not pci_bw_max:
+            ...
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:486]:
```suggestion
        if not pci_bw_max:
```

`...` was just a placeholder for code, we don't need it.

**Review by ilyam8 — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]

**Comment by ilyam8** [comment]
i tested the changes

**Pull request merged**